lib/handlers/evm/provider/InstrumentedEVMProvider.ts

```diff
@@ -0,0 +1,311 @@
+import { ethers } from 'ethers'
+import { Deferrable } from '@ethersproject/properties'
+import { Listener, TransactionRequest } from '@ethersproject/providers'
+import {
+  Block,
+  BlockTag,
+  BlockWithTransactions,
+  EventType,
+  Filter,
+  Log,
+  TransactionReceipt,
+  TransactionResponse,
+} from '@ethersproject/abstract-provider'
+import { metric, MetricLoggerUnit } from '@uniswap/smart-order-router'
+import { BigNumber, BigNumberish } from '@ethersproject/bignumber'
+import { Network, Networkish } from '@ethersproject/networks'
+import { ConnectionInfo } from '@ethersproject/web'
+import { ProviderName } from './ProviderName'
+
+export type InstrumentedEVMProviderProps = {
+  url?: ConnectionInfo | string
+  network?: Networkish
+  name: ProviderName
+}
+
+export class InstrumentedEVMProvider extends ethers.providers.JsonRpcProvider {
+  private readonly name: ProviderName
+  private readonly metricPrefix: string
+
+  constructor({ url, network, name }: InstrumentedEVMProviderProps) {
+    super(url, network)
+    this.name = name
+    this.metricPrefix = `RPC_${this.name}_${this.network.name}`
+  }
+
+  override call(transaction: Deferrable<TransactionRequest>, blockTag?: BlockTag | Promise<BlockTag>): Promise<string> {
+    metric.putMetric(`${this.metricPrefix}_CALL_REQUESTED`, 1, MetricLoggerUnit.Count)
```

Comment:
Could we change things to:

```
return super.call(transaction, blockTag)
  .then((response) => {
     metric.putMetric(`${this.metricPrefix}_CALL_SUCCESS`, 1, MetricLoggerUnit.Count)
     return response
   })
   .catch((error) => {
      metric.putMetric(`${this.metricPrefix}_CALL_FAILURE`, 1, MetricLoggerUnit.Count)
      throw error
  })
  .finally(() =>
    metric.putMetric(`${this.metricPrefix}_CALL_REQUESTED`, 1, MetricLoggerUnit.Count)
  )
```

That way we can see the request metrics alongside the other metrics (consider that the call can take several seconds)

Reply:
Sure